R/derive_extreme_event.R

```diff
@@ -0,0 +1,205 @@
+#'       condition = PARAMCD == "WAKE UP" & AVALC == "Y",
+#'       set_values_to = exprs(AVALC = "Waking up more than three times", AVAL = 2)
+#'     ),
+#'     event(
+#'       condition = PARAMCD == "FALL ASLEEP" & AVALC == "Y",
+#'       set_values_to = exprs(AVALC = "More than 30 mins to fall asleep", AVAL = 3)
+#'     ),
+#'     event(
+#'       condition = all(AVALC == "N"),
+#'       set_values_to = exprs(
+#'         AVALC = "No sleeping problems", AVAL = 4
+#'       )
+#'     ),
+#'     event(
+#'       condition = TRUE,
+#'       set_values_to = exprs(AVALC = "Missing", AVAL = 99)
+#'     )
+#'   ),
+#'   order = exprs(ADY),
+#'   mode = "last",
+#'   set_values_to = exprs(
+#'     PARAMCD = "WSP",
+#'     PARAM = "Worst Sleeping Problems"
+#'   )
+#' )
+derive_extreme_event <- function(dataset,
+                                 by_vars = NULL,
+                                 events,
+                                 order,
+                                 mode,
+                                 check_type = "warning",
+                                 set_values_to) {
+  # Check input parameters
+  assert_vars(by_vars, optional = TRUE)
+  assert_list_of(events, "event")
+  assert_order_vars(order)
+  assert_data_frame(
+    dataset,
+    required_vars = by_vars
+  )
+  mode <- assert_character_scalar(mode, values = c("first", "last"), case_sensitive = FALSE)
+  check_type <-
+    assert_character_scalar(
+      check_type,
+      values = c("none", "warning", "error"),
+      case_sensitive = FALSE
+    )
+  assert_varval_list(set_values_to)
+
+  # Create new observations
+  ## Create a dataset (selected_records) from `events`
+  condition_ls <- map(events, "condition")
+  set_values_to_ls <- map(events, "set_values_to")
+  event_order <- map(seq_len(length(events)), function(x) x)
+
+  selected_records_ls <- purrr::pmap(
+    list(condition_ls, set_values_to_ls, event_order),
+    function(x, y, z) {
+      dataset %>%
+        group_by(!!!by_vars) %>%
+        filter(!!enexpr(x)) %>%
+        mutate(!!!y, event_no = z)
+    }
+  )
+  selected_records <- bind_rows(selected_records_ls)
+
+  if (check_type != "none") {
+    signal_duplicate_records(
+      selected_records,
+      by_vars = exprs(event_no, !!!by_vars, !!!extract_vars(order)),
+      msg = paste0(
+        "Event dataset", " contains duplicate records with respect to ",
+        enumerate(vars2chr(exprs(!!!by_vars, !!!extract_vars(order), event_no)))
+      ),
+      cnd_type = check_type
+    )
+  }
+
+  ## Filter extreme within each event
+  selected_records_extreme <- selected_records %>%
+    group_by(event_no, !!!by_vars) %>%
+    arrange(!!!order, .by_group = TRUE) %>%
+    slice(if_else(mode == "first", 1L, n())) %>%
+    ungroup()
+
+  ## Filter extreme among all events
+  new_obs <- selected_records_extreme %>%
+    filter_extreme(
+      by_vars = by_vars,
+      order = exprs(event_no),
+      mode = "first",
+      check_type = check_type
+    ) %>%
```

Comment:
Do we need to split the selection? I wonder if we could do it with one call. E.g.,
```
if (mode == "first") {
  event_no_expr <- expr(event_no)
} else {
  event_no_expr <- expr(desc(event_no))
}
new_obs <- selected_records_extreme %>%
    filter_extreme(
      by_vars = by_vars,
      order = exprs_c(event_no_expr, order),
      mode = mode,
      check_type = check_type
    ) 
```

Reply:
should `mode` be applied to `order` instead of `event_no`? in the above example, `mode=first` or `last` will return the same results, e.g., `first` - order by event 1,2,3 and choose the first one (event_no=1); `last` - order by event 3,2,1 and choose the last one (event_no=1).
if we combine the two calls, maybe we can fix `mode="first"` in `filter_extreme`, but I don't know how to "reverse" the `order` (like above example for `event_no`) when it contains multiple variables

Reply:
I incorporated the `order` info to a temporary obs number `tmp_obs`  (by `by_vars` and `tmp_event_no`) as a workaround

Reply:
Your approach works as well.